Add module dependencies required to run SwitchYard quickstarts

igarashitm opened the pull request.

tdiesler commented:
Again, this PR does not does not pass basic validation - please run 'mvn clean install'.

To successfully add a module, it must be defined in the [dependencyManagement](https://github.com/tdiesler/wildfly-camel/blob/master/pom.xml#L129) section and be added [here](https://github.com/wildfly-extras/wildfly-camel/blob/master/modules/pom.xml#L50)

tdiesler commented on modules/etc/smartics/camel-modules.xml:
```diff
@@ -433,6 +437,38 @@
             <module name="javax.api"/>
             <module name="org.osgi.core"/>
             <module name="org.slf4j"/>
+            <module name="rhino" export="true">
```
Is it really true that component mqtt needs to see rhino, python, jruby, groovy. I don't see these dependencies [here](https://github.com/apache/camel/blob/master/components/camel-mqtt/pom.xml)

igarashitm replied:
Hi @tdiesler ,

It's **camel-script**, not camel-mqtt.

tdiesler replied:
Sorry, I miss read the diff. When this branch passes we can tell more

igarashitm commented:
OK, was my bad, sorry, I've been building it with -DskipTests=true. Now I saw 2 errors in itests/standalone:

* weatherIntegrationTest : Caused by: org.yaml.snakeyaml.error.YAMLException: Class not found: co.freeside.betamax.tape.yaml.YamlTape
* org.wildfly.camel.test.classloading.ExportedPathsTest.testExportedPaths
java.lang.AssertionError: Matching path: antlr/ASdebug

I have no idea why my change causes this failure for now. I'll look into further.

tdiesler commented:
The ExportedPathsTest is a sanity check that we don't expose more paths to the client than necessary. Every exposed path becomes public API and we would like this to remain as small as possible.

igarashitm commented:
Then how can I add those scripts as a public API? It's same thing that bsh is already exposed there. Just adding another script languages to camel-script.

tdiesler commented:
bsh is an integral part of the camel-script module (i.e. not a separate module). We choose which paths we want to expose for bsh like this

```<exports>
            <exclude path="bsh**" />
        </exports>
```

igarashitm commented:
OK got it, so it should be fine if I contain those jars for additional scripts into camel-script module itself and use exclude tag just like bsh then?

tdiesler commented:
yes, a good indicator is whether the test case (i.e. a client) has a package import for a given 3rd party script path. It not, visibility should likely be private for the camel-script component

igarashitm commented:
Thanks for the guide Thomas! All we need is just META-INF/services for each scripts to lookup script engine services, so that should be fine. I'll make that update.

tdiesler commented:
If you say "we need" - who is we? Is it the camel-script component, switchyard internals or some client code? IMHO, it should be sufficient for camel-script to see/load these services (i.e. this is an impl detail of camel-script)

igarashitm commented:
I meant we=SwitchYard - It's for SwitchYard quickstart applications which use those scripts in camel route. SwitchYard hit script engine not found error if we don't have these dependencies.

tdiesler commented:
Lets see a stack trace for this. I'd like to see who exactly needs to see these service paths. Still, IMHO this should be handled by camel-script and not leak into higher levels

igarashitm commented:
I can reproduce it and get stacktrace later, but that's why wildfly-camel already exporting META-INF of beanshell, isn't it? It just excludes bsh**, so META-INF is exported.

tdiesler commented:
This might be an oversight and not a requirement from the client - let me check that ...

tdiesler commented:
Related to #501, #500, #499, #498

igarashitm commented:
Updated with "push -f", removed rhino/groovy/python/jruby modules and added those jars to camel-script module. mvn clean install passed here. Added auto-generated stuffs as well.

I'll try excluding META-INF** in camel-script and see if it fails or not in SwitchYard distribution tests.

igarashitm commented:
Excluding META-INF** in camel-script causes following error:
http://pastebin.com/Fh9bFyBj

tdiesler commented:
This is the relevant part

```
Caused by: org.apache.camel.NoSuchLanguageException: No language could be found for: js
        at org.apache.camel.impl.DefaultLanguageResolver.noSpecificLanguageFound(DefaultLanguageResolver.java:101)
        at org.apache.camel.impl.DefaultLanguageResolver.resolveLanguage(DefaultLanguageResolver.java:80)
        at org.apache.camel.impl.DefaultCamelContext.resolveLanguage(DefaultCamelContext.java:1898)
```

We need to figure out what ClassLoader the DefaultLanguageResolver uses. If it works with the deployment's CL we can bring it in as is and discuss language resolution in the context of a separate  issue

tdiesler commented:
Processed by #506 because of #507
